pkgs/applications/science/math/mathematica/default.nix

```diff
@@ -0,0 +1,147 @@
+{ stdenv
+, coreutils
+, patchelf
+, callPackage
+, alsaLib
+, dbus
+, fontconfig
+, freetype
+, gcc
+, glib
+, ncurses
+, opencv
+, openssl
+, unixODBC
+, xkeyboard_config
+, xorg
+, zlib
+, libxml2
+, libuuid
+, lang ? "en"
+, libGL
+, libGLU
+}:
+
+let
+  l10n =
+    with stdenv.lib;
+    with callPackage ./l10ns.nix {};
+    flip (findFirst (l: l.lang == lang)) l10ns
+      (throw "Language '${lang}' not supported");
+in
+stdenv.mkDerivation rec {
+  inherit (l10n) version name src;
+
+  buildInputs = [
+    coreutils
+    patchelf
+    alsaLib
+    coreutils
+    dbus
+    fontconfig
+    freetype
+    gcc.cc
+    gcc.libc
+    glib
+    ncurses
+    opencv
+    openssl
+    unixODBC
+    xkeyboard_config
+    libxml2
+    libuuid
+    zlib
+    libGL
+    libGLU
+  ] ++ (with xorg; [
+    libX11
+    libXext
+    libXtst
+    libXi
+    libXmu
+    libXrender
+    libxcb
+    libXcursor
+    libXfixes
+    libXrandr
+    libICE
+    libSM
+  ]);
+
+  ldpath = stdenv.lib.makeLibraryPath buildInputs
+    + stdenv.lib.optionalString (stdenv.hostPlatform.system == "x86_64-linux")
+      (":" + stdenv.lib.makeSearchPathOutput "lib" "lib64" buildInputs);
+
+  phases = "unpackPhase installPhase fixupPhase";
+
+  unpackPhase = ''
+    echo "=== Extracting makeself archive ==="
+    # find offset from file
+    offset=$(${stdenv.shell} -c "$(grep -axm1 -e 'offset=.*' $src); echo \$offset" $src)
+    dd if="$src" ibs=$offset skip=1 | tar -xf -
+    cd Unix
+  '';
+
+  installPhase = ''
+    cd Installer
+    # don't restrict PATH, that has already been done
+    sed -i -e 's/^PATH=/# PATH=/' MathInstaller
+    sed -i -e 's/\/bin\/bash/\/bin\/sh/' MathInstaller
+
+    echo "=== Running MathInstaller ==="
+    ./MathInstaller -auto -createdir=y -execdir=$out/bin -targetdir=$out/libexec/Mathematica -silent
+
+    # Fix library paths
+    cd $out/libexec/Mathematica/Executables
+    for path in mathematica MathKernel Mathematica WolframKernel wolfram math; do
+      sed -i -e 's#export LD_LIBRARY_PATH$#export LD_LIBRARY_PATH=${zlib}/lib:\''${LD_LIBRARY_PATH}#' $path
+    done
+
+    # Fix xkeyboard config path for Qt
+    for path in mathematica Mathematica; do
+      sed -i -e "2iexport QT_XKB_CONFIG_ROOT=\"${xkeyboard_config}/share/X11/xkb\"\n" $path
+    done
+  '';
+
+  preFixup = ''
```

Comment:
Can we replace this to use `autoPatchelfHook` instead?

Reply:
I got this error when I tried to resolve this issue:

```
searching for dependencies of /nix/store/wamni5hkz0ic30wn41gpf5p947lhdgja-mathematica-11.3.0/libexec/Mathematica/SystemFiles/Libraries/Linux-x86-64/libmkl_ao_worker.so                                                                                                  
  libcoi_device.so.0 -> not found!                                                                                                  
  libiomp5.so -> found: /nix/store/wamni5hkz0ic30wn41gpf5p947lhdgja-mathematica-11.3.0/libexec/Mathematica/SystemFiles/Libraries/Linux-x86-64/libiomp5.so
```

After doing some search, I found that `libcoi_device.so.0` was part of Intel Manycore Platform Software Stack. It can be downloaded from [here](https://software.intel.com/en-us/articles/intel-manycore-platform-software-stack-mpss#lx38rel).

Do we introduce such a set of packages for Mathematica? I'm not sure how many to add yet.

Reply:
To elaborate on the last comment, here are the observations after some investigations:

1. The source code of MPSS is broken down into many modules, with `mpss-coi` being one of them, and some modules depend on others. To successfully build `mpss-coi`, which Mathematica requires, it is necessary to install `mpss-metadata` and `libscif` at least. I have not yet successfully built `mpss-coi`, so I'm unsure what the whole list contains.

1. It is said in the document that at least one Intel(R) Xeon Phi(TM) coprocessor is necessary to install MPSS, so even if I manage to build `mpss-coi` successfully, I will not be able to verify if the result is working (I don't have such hardware).

Here are some proposals:

1. Keep the good old `installPhase`, because it works.

1. Use `autoPatchelfHook`, but do it manually (i.e., setting `dontAutoPatchelf` to `true`), and keep files that require such complex dependencies temporarily out. Or better, add some options for future updates.

1. Package the dependencies anyway, and find somebody verifying that the result is working. There are some uncertainties here.

I personally prefer the second proposal, because it requires minimal effort to provide a working installation, and makes things left undone clear. How do you think?

Reply:
I think the second option is the best. I think we can just delete the rogue library before the `autoPatchelfHook` is executed.

Reply:
I have fixed that issue, but now `patchelf` needs to be upgraded to version 0.10 to finish the build. That introduces many rebuilds of packages, including `gcc`, so it will take a while. Let's see the outcome later.

Reply:
What is the issue with patchelf 0.9? I think you could just add `patchelfUnstable` to `nativeBuildInputs` and it should shadow the older patchelf from `stdenv`.

Reply:
An error like this with `patchelf` 0.9:

```
$ /nix/store/s7p0wfbdfbq307zgifhnh6w8sfjvy64y-patchelf-0.9/bin/patchelf --set-interpreter /nix/store/avr2x43njlq4kyb1a9zgrh6fih5fq4si-glibc-2.27/lib/ld-linux-x86-64.so.2 --debug /nix/store/b6zk8in4n340qvmrs54jan0pbgzfqygs-mathematica-11.3.0/libexec/Mathematica/SystemFiles/Components/SemanticImport/Binaries/Linux-x86-64/server
patching ELF file `/nix/store/b6zk8in4n340qvmrs54jan0pbgzfqygs-mathematica-11.3.0/libexec/Mathematica/SystemFiles/Components/SemanticImport/Binaries/Linux-x86-64/server'
Kernel page size is 4096 bytes
replacing section `.interp' with size 80
this is an executable
using replaced section `.interp'
last replaced is 3
looking at section `.tbss'
replacing section `.tbss' which is in the way
looking at section `.note.go.buildid'
replacing section `.note.go.buildid' which is in the way
looking at section `.interp'
first reserved offset/addr is 0x1000/0x401000
first page is 0x400000
needed space is 768
clearing first 3472 bytes
cannot find section .tbss
$ echo $?
1
```

Replacing `patchelf` to 0.10 resolves the above issue.

Reply:
Okay, adding `patchelfUnstable` to `nativeBuildInputs` also fixes the issue. I'll use this one.

When would it be appropriate to update `patchelf` to 0.10? After it being more widely tested, I guess?

Reply:
It appears there some work needed to be done before `patchelf` version can be bumped. See #58715

Reply:
Still more errors to fix. It's late night here and I'll continue the work tomorrow.